### ISSUE TITLE
Abnormal memory consumption in SimCode with the NB

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10231]
I recently added [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) to the [LargeTestSuite_NB](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/LargeTestSuite_NB.html) job.

The system has 1.6 million variables, which are all states. Unfortunately the test [fails](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err), I guess because the allocated 60 GB are not enough.

Looking at the two smaller cases, [HarmonicOscillator_N_204800](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_204800.err) allocates 12 GB of memory for the SimCode phase, and [HarmonicOscillator_N_409600](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.err) almost 25 GB. The scaling is not superlinear, which is good, but the code is probably quite inefficient memory-wise. 

Consider that N = 409600 case has 0.8 M variables and 1.6 M non-zero entries in the Jacobian, each equation or non-zero Jacobian entry consumes about 10 kB of RAM. I'm not sure why, but I guess this amount of memory could probably be reduced by at least one order of magnitude without too much effort. Maybe more. 

Unfortunately, even though the code to solve the equations is vectorized, the variables and the sparsity pattern in the generated code are still fully scalarized, and so is the init.xml file. Maybe the MetaModelica code that handles that is not very memory efficient. 

This is essential to break the 1M state limit and head towards the 10M goal.

**Comment by casella** [comment]
@perost I'd like to also have your opinion, regarding the efficiency of MetaModelica for SimCode. Why do we consume so much memory (about 10KB) for each individual variable? Could there be some fundamental improvement to make this more efficient?

**Comment by casella** [comment]
See @adrpo's comment in [#10400](https://github.com/OpenModelica/OpenModelica/issues/10400#issuecomment-1469724200)

**Comment by casella** [comment]
[HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) now works, despite SimCode allocating a whooping 50 GB of memory. Apparently, the garbage collector can recover some of it.

I'll try an even larger model to see what happens. In any case, even if this memory can be recovered, all these allocations take time, so we should try to make the process more efficient.

**Comment by casella** [comment]
In the meantime, the memory allocation of SimCode for [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) was slightly reduced from 50 to 43 GB. Not much, but at least it's not growing 😅

**Comment by casella** [comment]
I also noticed that, despite the memory usage of SimCode growing as O(N), the run time of SimCode grows as O(N^1.7) for models with more than 100,000 states. In principle, I see no reason why this should happen, but maybe I'm missing something.

@perost, could you please do some profiling of the SimCode phase for the larger LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_XXX models and figure out where the bottleneck is?

**Comment by perost** [comment]
I found that a lot of the time was spent in `SimCodeUtil.createVarToArrayIndexMapping`, which is part of the OB but still used by the NB. It created two hash tables of size 25343 and then shoved 3276803 elements into both of them, which obviously caused some performance issues.

 #13985 rewrites the function so that it takes into account the number of variables when creating the hash tables. This approximately halves the time for the SimCode phase for the HarmonicOscillator_N_819200 model.

The remaining time now seems to be mostly related to the sheer size of these tables, i.e. hard to optimize further without somehow reducing the number of variables we need to store in them.

**Comment by perost** [comment]
Actually, I made a copy-paste mistake which caused the state variables to be added twice, which fortunately broke some tests. With that fixed the time is now more than halved, from about 80s to 30s (40s with the mistake).

**Comment by casella** [comment]
Great!

So, how many hash tables and how many elements do we have now?

**Comment by perost** [comment]
> So, how many hash tables and how many elements do we have now?

Same as before, the change is just to make them appropriately sized to how many variables there are. The load factor of a hash table (elements / buckets) should ideally be a bit lower than 1, in this case it was over 100 which causes severe scaling issues.

**Comment by casella** [comment]
> The load factor of a hash table (elements / buckets) should ideally be a bit lower than 1, in this case it was over 100 which causes severe scaling issues.

I'm not sure if these scaling issues are paid in terms of memory or runtime (or both), but I guess this should be fixed, right?

The [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) model has 1,638,400 variables, which are all states, plus three scalar parameters. As I understand, the total of 3,276,803 elements corresponds exactly to the 1,638,400 variables, plus their 1,638,400 derivatives, plus the three parameters.

The question is, why are the hash tables originally sized to 25343 elements? Is that some "large enough" number that was OK for reasonably-sized old-fashioned models but clearly undersized for large ones? Or is there some criterion in determining this number?

@kabdelhak, @phannebohm you feedback would be precious here.

**Comment by perost** [comment]
> The question is, why are the hash tables originally sized to 25343 elements? Is that some "large enough" number that was OK for reasonably-sized old-fashioned models but clearly undersized for large ones? Or is there some criterion in determining this number?

It's one of the predefined sizes, basically just an arbitrary large prime number:
https://github.com/OpenModelica/OpenModelica/blob/85dbf75b79c578479f423867cfe6836a5848b5f7/OMCompiler/Compiler/Util/BaseHashTable.mo#L79

The issue with the old hash tables we have are that they have a fixed size, so if you don't know in advance how many elements will be added you just have to choose a size and pray. Needless to say this has been a major source of scaling issues. In this case we do know how many elements will be added though, the code just didn't take advantage of it.

The new hash tables I implemented a few years ago avoid this issue by resizing when the load factor becomes too large, but we still have a lot of old code that hasn't been rewritten to use the new hashtables yet.

**Comment by casella** [comment]
Do you think you can rewrite this specific piece of code with the new hashtables?

**Comment by perost** [comment]
> Do you think you can rewrite this specific piece of code with the new hashtables?

I'm not sure, they're used in the code generation templates and I don't know if Susan can handle them. And it's not really needed at the moment since the scaling issue is fixed now.

**Comment by casella** [comment]
OK, so #13985 should have fixed the scaling issue for good.

**Comment by perost** [comment]
> OK, so [#13985](https://github.com/OpenModelica/OpenModelica/pull/13985) should have fixed the scaling issue for good.

Yes

**Comment by casella** [comment]
Good. I scheduled a [LargeTestSuite](https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/9541/console) test run, as soon as the pipeline is unclogged we'll see what the results are on the CI, also for other models. Thanks!

**Comment by perost** [comment]
> Good. I scheduled a [LargeTestSuite](https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/9541/console) test run, as soon as the pipeline is unclogged we'll see what the results are on the CI, also for other models. Thanks!

The [results are in](https://libraries.openmodelica.org/branches/history/heavy_tests/2025-01-25%2017:59:16..2025-06-11%2001:35:35.html), but it seems it was almost half a year since the last run so it's a bit messy. But it seems all of the models now scale approximately linear in the SimCode phase now at least.

**Comment by casella** [comment]
> The [results are in](https://libraries.openmodelica.org/branches/history/heavy_tests/2025-01-25%2017:59:16..2025-06-11%2001:35:35.html), but it seems it was almost half a year since the last run so it's a bit messy. But it seems all of the models now scale approximately linear in the SimCode phase now at least.

Yes, indeed we should have run it more often. There is a critical regression on the C compile time, see  #13999, but that is downstream.

The run time of SimCode is now growing linearly, rather than scaling as O(N^1.7) for the largest models as I reported [last week](https://github.com/OpenModelica/OpenModelica/issues/10231#issuecomment-2945367849). Also, the memory allocation for SimCode in [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) went down from the 43 GB I reported [one year ago](https://github.com/OpenModelica/OpenModelica/issues/10231#issuecomment-2162669261) to 16 GB. 

It's still a lot, so there are probably other possible optimization, but I'd say that a factor 3 memory reduction and getting to O(N) time is a major result, so I'll close this ticket.

Thanks @perost!
